---
title: "coi5p-vignette"
author: "Cameron M. Nugent"
data: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{coi5p-vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
 
```{r setup}
library(coi5p)
```


## Abstract

**coi5p** is an R package designed for the cleaning, contextualization and assessment of cytochrome c oxidase I DNA barcode data (COI-5P -- *link and brief definition* ). It contains functions for placing COI-5P barcode sequences into a common reading frame, translating DNA sequences to amino acids and for assessing the likelihood that a given barcode sequence includes an insertion or deletion error. These functions are provided as a single function analysis pipeline and are also avaliable individually for efficient and targeted analysis of barcode data.

## Introduction

The backbone of the coi5p package is a pair of profile hidden Markov models (PHMMs) that have been trained using a representitive sample of the COI-5P sequences avalaible on [the BOLD database](http://www.boldsystems.org/index.php). A 657 nucleotide PHMM receives raw sequences from the user and uses the Viterbi algorithm (implemented via [the R package aphid](https://github.com/shaunpwilkinson/aphid)) to optimally match the input sequence against the COI-5P nucleotide profile. The second PHMM receives an amino acid sequence that is matched against the COI-5P amino acids profile. The model provides two boolean output metrics to the user: does the sequence contain stop codons and is the sequence likely to contain an insertion or deletion error. The indel boolean is based on the log likelihood of the amino acid sequence compared to the PHMM. A default indel likelihood threshold of -346.95 is set (see [paper]() for basis of this threshold). Sequences with likelihood values less than this threshold indicate the sequence is likely to contain an indel error, as the amino acid sequence is improbable and therefore indicative of a frame shift.

The nucelotide and amino acid PHMMs are interfaced through the translate function, which takes the in frame nucleotide sequence and translates it to amino acids. This function uses [the sequinr package]() to conduct translation in all instances where [the genetic code associated with the sample](https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi) is known. For samples without taxonomic IDs or known genetic codes an additional genetic code is provided. This genetic code is used to conduct censored translation, meaning that translation is conducted normally for codons that do not vary in the amino acid they code for across all known mitochondrial genetic codes. The codons that are known to vary in the amino acid they code for across taxa are not translated, rather a placeholder (`?`) is output to indicate that the amino acid at this location in the sequence cannot be stated with certainty. This functionality allos the indel check function to assess the likelihood of sequences of unknown taxonomy without being overly stringent in its characterization of sequencs as indels due to the appropriation of the wrong translation table.

The translation table employed in censored translation - five codons are translated to placeholder question marks, due to their ambiguity across different mitochondrial translation tables.
```
 Censored translation table:
            FFLLSSSSYY?*CCWWLLLLPPPPHHQQRRRRII?MTTTTNN?KSS??VVVVAAAADDEEGGGG
   Base1  = TTTTTTTTTTTTTTTTCCCCCCCCCCCCCCCCAAAAAAAAAAAAAAAAGGGGGGGGGGGGGGGG
   Base2  = TTTTCCCCAAAAGGGGTTTTCCCCAAAAGGGGTTTTCCCCAAAAGGGGTTTTCCCCAAAAGGGG
   Base3  = TCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAG
```

## The coi5p package
### Dependencies

### Data structure


### Using the package
#### Full analysis pipeline for a single sequence
An example execution of the complete coi5p analysis pipeline with default options is demonstrated below using an example COI-5P barcode DNA sequence.

```{r}
example_nt_string = 'ctctacttgatttttggtgcatgagcaggaatagttggaatagctttaagtttactaattcgcgctgaactaggtcaacccggatctcttttaggggatgatcagatttataatgtgatcgtaaccgcccatgcctttgtaataatcttttttatggttatacctgtaataattggtggctttggcaattgacttgttcctttaataattggtgcaccagatatagcattccctcgaataaataatataagtttctggcttcttcctccttcgttcttacttctcctggcctccgcaggagtagaagctggagcaggaaccggatgaactgtatatcctcctttagcaggtaatttagcacatgctggcccctctgttgatttagccatcttttcccttcatttggccggtatctcatcaattttagcctctattaattttattacaactattattaatataaaacccccaactatttctcaatatcaaacaccattatttgtttgatctattcttatcaccactgttcttctactccttgctctccctgttcttgcagccggaattacaatattattaacagaccgcaacctcaacactacattctttgaccccgcagggggaggggacccaattctctatcaacactta'
example_nt_string


output = coi5p_pipe(example_nt_string)
output
```
Executing the entire pipeline yields a coi5p object. Calling the name of the object variable prints the coi5p object's summary and shows us all of the important information, including the original raw sequence, the sequence set in reading frame, the amino acid sequence and the summary stats regarding the likelihood of the sequence containing an error.

We can obtain the individual components of the object using the dollar sign notation.
```{r}
#see the avaliable components
names(output)
#retrieve only the framed sequence from the object
output$framed
```

#### Calling functions individually

#### Example of a batch analysis of barcode sequences
Here we will be working with the example dataframe (`example_barcode_data`) 

Although sequence data input/output is outside of the scope of the coi5p package, an example of how to load a fasta file into a dataframe using the seqinr package is inlcuded in the supplementary section at the end of this vignette. 

# work with the csv data stored in the /data/ folder
```{r}
#this is the example data 
head(example_barcode_data)

```


### Supplementary Information
#### Loading and manipulating a fasta file
```{r}
library(seqinr)
# example fasta file included with coi5p
ex_fasta_file = system.file("extdata/example_barcode_data.fasta", package = "coi5p")

ex_data = seqinr::read.fasta(ex_fasta_file)
#then parse to the column setup I have


```



# TODO - add a fasta file to the /data/ folder that contains ~10 sequences that cover a range of sequence types

-----------------------------------------------------------------------------------------------
#Once completed can - submit it to the Journal of Statistical Software or The R Journal. Online, R journals that are peer reviewed. The BOLD special issue is the first option.


This of this from a beginners perspective, use simple and accesible language as may be used by international researchers. goal is clear and accesible.
 - Modify this to explain/showcase the code and how it works
 - use `ctrl-shift-k` to knit the vignette and preview the output.
 
What this file will need to include:
 - a basic example of getting a string loaded into coi5p
 - how the different functions of the package are run, what they result in
 - explination as to why the functions are split - save user time b/c if you just want it in frame and not translated, that is possible!
 - show how this could be integrated into a workflow.
 



```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

